### ISSUE TITLE
v10 testing

### ISSUE THREAD

**Issue by will-stone** [will-stone/browserosaurus#200]
@cocoonkid @caterino @stgarf @VandrerenErik @aaronkim98 @kopach @optimusdag @CedricGatay @pdxbmw @chrisands @sushione @joel-tonuri @marcusramberg @Miracle7days @boscovn @Victorio115 @israelroldan @brettgoss @idontlie @thewoolleyman @5knnbdwm @lucagiovagnoli @kemayo @prewk @Kulcanhez @RyanRoberts @ngerasimatos @armandocin @mindfulsource @ibanezda @houndtt @rushiagr @SeeHavens @pook52mb @dimaboyko @drynda @aakside @Kilian @i0ntempest @sputnik13 @ethan-artos

Hi all,

Firstly, apologies for the mass @ tagging; I didn't really know the best way to approach this.

Over the past few months, I have been thinking about making some changes to Browserosaurus, and whilst I am currently between jobs, I decided now would be a good time to try out some updates. Below is a link to an alpha version (but essentially complete) upgrade: v10.0
I have tagged you all because you have recently logged issues or PRs and I would love it if you could give v10 a go and let me know what you think on this thread. I won't say anything about any of the changes now as I want it to be used without preconceptions, but I will happily discuss design decisions once you've tried it out.

https://www.dropbox.com/s/tr7m5450s5e4k05/Browserosaurus-10.0.0-alpha.3.dmg?dl=0

I will remove this link after a week or so as I'm not sure if DropBox is the best place for it.

Thanks and I look forward to hearing your feedback,

Will.

**Comment by optimusdag** [comment]
Right out of the box it works great!
Just a request... can you add COPY option to be able to be the default/favorited/starred one? That's the one I use the most.

**Comment by optimusdag** [comment]
BTW... me being so rude... thanks for all the effort! This is a really useful app and I love it...

**Comment by kopach** [comment]
Awesome! Thank you.

**Comment by Kulcanhez** [comment]
I've been using it for the last 10 minutes. Thank you for your amazing work. ❤️ 
I love the ability to choose my favorite browser! I like the new window but I prefer the old window for when I click in the links and the new window for when I click in the app icon and want customization. I would also love to have one key shortcut to copy the link, for example, the dot or something else. Thank you once again!

**Comment by pdxbmw** [comment]
Initial impressions are quite good. Thank you, Will!

I was never able to figure out how to solve the issue with multiple windows that crept up with Electron v8. I spent probably 10-20 hours troubleshooting and working on a solution, but nothing fully worked. I even went as far to see if I could help you port over to a native app in Xcode, but ran out of time.

All that’s to say, thank you for continuing to develop and improve this app. It’s greatly appreciated. I ❤️ Browserosaurus.

**Comment by joel-tonuri** [comment]
Absolutely awesome, Will! Appreciate you taking such a great deal of time to work on this!

Thank you once again for this absolutely splendid app!

**Comment by CedricGatay** [comment]
The new UI is really great ! It is rather disturbing to have the popup in the middle of the main screen even if we're clicking on another but I bet it's just an habit that will easily change.

I will test it more during my workdays !

**Comment by sushione** [comment]
Seems to be working very well now on virtual desktop ! A big thank you because your app is a must have !

I use to like a lot the minimalist design before but the new one is ok.
Just one thing, I think it was easier to have the chooser windows close to the mouse/link to open, because sometimes it is far away from the center of my screen (and I use two monitors) and my keyboard is also not close. Before I was able to choose the default browser with minimal mouse move.

Thank you Will you did a great job !

**Comment by 5knnbdwm** [comment]
Have to say it is working a lot better than v9. With v9 it regularly happened that I had to restart browserosaurus to get it to work again, as the popup window didn't appear or send me to a different screen and didn't show up either.

Other than that I have to agree with others, I really like the small minimalistic popup window. Don't need to configure the app in a normal workflow. That doesn't, however, mean that I don't like the design overall.

**Comment by will-stone** [comment]
Thanks everyone. Generally positive so far which is great!

> I think it was easier to have the chooser windows close to the mouse/link to open

I knew it wouldn't be long until this was mentioned but I thought I'd wait until it had to explain this design decision, so... the old B used a rather hacky (but interesting, even if I do say so myself) way of moving the picker-window next to the mouse: it was actually the whole window but most of it was translucent. It did this:

1. Click link.
2. B, currently hidden, moved to the display where the mouse is, and expanded/shrunk to the dimensions of the whole display.
3. The coordinates of the mouse were sent to the renderer process, where it would draw a box (the picker), and the blue bar at the top.
4. The app would then be shown.

This allowed me, (in v9 or v8, can't remember) to add the semi-translucent backdrop, which I thought helped accentuate the picker.

All this moving around and resizing of the window always felt wrong. For example, I had to keep track of when the mouse was over the picker or over the translucent background, so that I could turn on and off mouse capturing in case the user wanted to dismiss B and click through to something behind. We also had a long running issue with people using full-screen apps; Browserosaurus just can't display on top (like Spotlight/Alfred can) due to limitations with Electron. This meant that it would change back to another screen, fire the _blur_ event, and disappear before the user had seen the window.

Now we don't have all this programmatic window moving, using `setVisibleOnAllWorkspaces` (which we've has [mixed results](https://github.com/will-stone/browserosaurus/pull/184) with before) should now just work.

The downside to not moving the window has been noted by @CedricGatay:

> It is rather disturbing to have the popup in the middle of the main screen even if we're clicking on another but I bet it's just an habit that will easily change.

If you don't have "Displays have separate spaces" ticked, this is going to be the case. I'm currently on a one-screen setup at the moment, but I hope to be getting a monitor soon where I can test the day-to-day feel of B not moving, but I feel that the benefits of this approach outweigh the negatives.

---

I had a feature request recently that asked about changing the colours of the UI. For those interested, the grey palette comes from my VSCode theme [Plastic](https://will-stone.github.io/plastic/). I hope to develop a light theme for v10.1

---

@optimusdag I am intrigued by this request:

> Just a request... can you add COPY option to be able to be the default/favorited/starred one? 
That's the one I use the most.

I mapped `cmd + c` to copy as that's the system default. I was wondering, if you use copy that much that you'd want it mapped to `space`, what do you do next? Where do you copy the URL? I presume it's not a browser... unless it's one B doesn't currently support? 🤔

---

Lastly, do not fear if you would like to keep v9. After v10 is released, I will release a version of v9 that has the auto-updater removed. And the the auto-updater in v10 only notifies that there's a new version, it doesn't download it.

**Comment by optimusdag** [comment]
Hi Will,
Regarding your question. I only use Chrome, but I have several Chrome
"personas" open at the same time (each one with it's own window), one for
personal stuff, one for main work account, one for other freelance account,
etc. So I use Browserosaurus to avoid immediate opening of the last used
chrome window (that could be any of the personas), I copy the link and
manually choose the persona window I like to use. This way my navigation
history, youtube history, bitbucket account, etc do not get mixed between
personal, work1, work2, etc.
That's why I need to copy as my "favorite" action, unless you can come to a
solution where one can choose a different persona and not just the browser,
but I think that is too much to ask since I doubt that is a very common use
case.

Thanks!

Daniel Astudillo G.
Ingeniero, padre, músico, gamer, geek, deportista... vivo la vida con
actitud positiva
http://www.dag.cl
Perfil: http://about.me/optimusdag

**Comment by will-stone** [comment]
ah profiles 😬 We have [looked into](https://github.com/will-stone/browserosaurus/issues/56) that [in the past](https://github.com/will-stone/browserosaurus/pull/188). I was hoping for B to have a plugin system in place for v10 that could potentially allow someone to create and maintain their own solution to this but alas it was a little too ambitious this time around. Also, the UX was a nightmare. However, the new architecture may help make this a little easier to implement, I'm simply not a fan of putting in browser-specific features, but I'm not ruling it out completely.

**Comment by will-stone** [comment]
Okay peeps, here's alpha 4: https://www.dropbox.com/s/0momfpyx0vdury1/Browserosaurus-10.0.0-alpha.4.dmg?dl=0

- Only prompt to set as default browser on first run.
- Added button in status bar to set as default browser when it is detected that B is not... so you can ignore it if you'd rather keep another app as the default http client.
- Check for update on startup, this won't be testable with alpha releases but here's the difference to v9: v9 uses Electron-AutoUpdater which is a bit of a bulldozer and will always auto download and install on restart. I've written my own update checker: it will check for an update but only display a button in the status bar if it finds one. On clicking this, the URL bar will change to "http://browserosaurus.com" and the user can choose a browser to open it with as normal:

<img width="912" alt="image" src="https://user-images.githubusercontent.com/654103/83569589-a1c60d00-a51c-11ea-8636-a57192652d64.png">

- The history list will only keep the last 10 links: I don't want this to become a bookmark manager and expect people will save links in their usual place when they need to retain this information, but it is enough in case you quickly need a link you know you recently clicked. There is now a clear history button:

<img width="912" alt="image" src="https://user-images.githubusercontent.com/654103/83570739-7cd29980-a51e-11ea-8300-1e2676033416.png">

**Comment by sushione** [comment]
Oops, something wrong ! everything empty on chooser window !

**Comment by sushione** [comment]
Ok, back to normal, had to restart the app ! working great

**Comment by will-stone** [comment]
That was probably because of the extra key I added to the store to detect if it's the `firstRun`. Hopefully won't be a problem when users upgrade from v9 (as the stores have different names).

**Comment by chrisands** [comment]
Testing the new release for a few days and it is great! 
It feels much faster and the problem with workspaces doesn't appear in the new version.

I think that the bottom bar is possible to remove because it mostly doesn't do anything.
![Screenshot 2020-06-03 at 10 14 20](https://user-images.githubusercontent.com/12214583/83607282-9823c000-a583-11ea-808d-b14d660213f4.png)

**Comment by will-stone** [comment]
I did have everything in a burger menu at one point in the development. However, there’s a couple of issues with removing it:

- the status bar gives more space for the modal menus when you have 5 or less (one row of) browsers.
- buttons in buttons can be tricky and can be clicked accidentally. 
- the discoverability of staring a browser would be hidden which is not recommended UX.
- users may not be aware that setting as default browser is the recommended way to use B and hiding this button could further confuse.

So far I’m liking having the status bar as a way to extend functionality and stuffing all future features in a burger menu may not be fun. And just a note: unfortunately I cannot use the Apple menu bar menus as these are removed by the option that keeps B out of the dock and cmd-tab menu.

**Comment by CedricGatay** [comment]
It works very well this way and it feels way snappier than v9 when moving along the cursor position was at times feeling sluggish.

I have a potential improvement now that you're considering the UI, I use Chrome and Chrome Canary, and the "G" key is automatically bound to "Chrome". Are you considering shortcut customization in this v10 ?

Anyway this is a big release ! Congratulations for such amount of work !

**Comment by will-stone** [comment]
Thanks @CedricGatay, it has indeed been a lot of work. The hardcoded hotkeys are reserved for stable releases, as otherwise I'd run out of keys fast 😆 I really want to add in hotkey customisation and it is high on my list.

**Comment by pdxbmw** [comment]
Congrats, Will!

**Comment by prewk** [comment]
I love it! so snappy! I hope someone is able to crack the profile code, I switch a lot between private/work profiles in Chrome :)

**Comment by SeeHavens** [comment]
Thanks for the update. My biggest feature request if it is possible is an option to open a new link in the background and leave the current app or window in focus.



Hope all is well,
–Chris


iTypos? iSorry.
---------------------------------
Christopher Havens
see.havens@gmail.com

**Comment by will-stone** [comment]
> Thanks for the update. My biggest feature request if it is possible is an option to open a new link in the background and leave the current app or window in focus.

Hold `alt` ("option" for the mac purists) when clicking a browser tile or hitting a keyboard shortcut (NOT when clicking the link). It doesn't work with all browsers and the browser may need to already need to be running. All it does is [add the `-g` flag to the `open` command](https://github.com/will-stone/browserosaurus/blob/db60e8eaa52d78011c1ab00287d9f1adfa79f87a/src/main/main.ts#L144), if a browser doesn't support it then it won't work and you will need to make a feature request with the browsers' authors.

**Comment by SeeHavens** [comment]
Thanks for the quick feedback! I will definitely  give this a go.

Thanks for this software, I use it all the time.



Hope all is well,
–Chris


iTypos? iSorry.
---------------------------------
Christopher Havens
see.havens@gmail.com

**Comment by will-stone** [comment]
@CedricGatay v10.1.0 has just been released which adds hotkey customisation 🙂

**Comment by armandocin** [comment]
I've been using v10 for these 5 days and it works really well. 

I've noticed that a issue I had before has been solved: I work with 3 monitors on a mac book pro and sometimes when I clicked a link the browsers choice window opened on desktop, and if I had some program window opened on desktop, browserosaurus window immediately lost focus. So I couldn't open links unless I clicked space very quickly. Again this didn't happen every time.

Now this seems solved, or at least if the browserosaurus window opens on desktop it doesn't lose focus.

In conclusion I like it a lot. Great job and thank you!

**Comment by Kilian** [comment]
Took me a while to check this out but the new version looks super slick and feels faster too, great job! :)
